### PR TITLE
operator: optimizate predicate

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

during debug https://github.com/istio/istio/issues/41417, I found that `reflect.DeepEqual(oldIOP.Spec, newIOP.Spec)` always return false with following yaml:

```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
  namespace: istio-system
spec:
  meshConfig:
    defaultConfig:
      proxyMetadata: {}
```

this PR will add new count metric for `Reconcile` and optimizate predicate

### CODE REVIEW

File: operator/pkg/controller/istiocontrolplane/istiocontrolplane_controller.go [hunk 1]
```diff
@@ -182,11 +184,28 @@ var (
...
+			if !reflect.DeepEqual(oldIOP.Spec, newIOP.Spec) {
+				old, err := json.Marshal(oldIOP.Spec)
+				if err != nil {
+					return true
+				}
+				new, err := json.Marshal(newIOP.Spec)
+				if err != nil {
+					return true
+				}
+				if !bytes.Equal(old, new) {
```
Comment (hzxuzhonghu) [inline operator/pkg/controller/istiocontrolplane/istiocontrolplane_controller.go:201]: Why reflect.DeepEqual(oldIOP.Spec, newIOP.Spec) always return false?

Comment (zirain) [reply operator/pkg/controller/istiocontrolplane/istiocontrolplane_controller.go]: the following yaml will cause this:
```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio
  namespace: istio-system
spec:
  meshConfig:
    defaultConfig:
      proxyMetadata: {}
```

**Review by hzxuzhonghu — APPROVED** [review]
LGTM, k8s will auto increase generation when a resource spec updated

**Review by howardjohn — COMMENTED** [review]
When does generation change on CRD?

**Comment by hzxuzhonghu** [comment]
@howardjohn  staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L147

**Review by howardjohn — APPROVED** [review]

**Pull request merged**